tasty/HasCal/Test/AsyncInterface.hs

```diff
@@ -0,0 +1,131 @@
+{-| This is the "Our second specification of an asynchronous interface" example
+    from p. 30 in "Specifying Systems" by Lamport:
+
+@
+        CONSTANT Data
+        VARIABLE chan
+        TypeInvariant := chan \in [val : Data, rdy : {0,1}, ack : {0, 1}]
+
+        Init := /\ TypeInvariant
+                /\ chan.ack = chan.rdy
+
+        Send(d) := /\ chan.rdy = chan.ack
+                   /\ chan' = [chan EXCEPT !.val = d, !.rdy = 1 - @]
+
+        Rcv := /\ chan.rdy != chan.ack
+               /\ chan' = [chan EXCPT !.ack = 1 - @]
+
+        Next := (\E d \in Data : Send(d)) \/ Rcv
+
+        Spec := Init /\ [][Next]_chan
+@
+
+    The following module is an attempt to translate it to HasCal.
+-}
+
+{-# LANGUAGE BlockArguments  #-}
+{-# LANGUAGE DeriveAnyClass  #-}
+{-# LANGUAGE DeriveGeneric   #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module HasCal.Test.AsyncInterface where
+
+import HasCal
+import Prelude hiding (either, init)
+import qualified Test.QuickCheck as QC
+import qualified Test.QuickCheck.Monadic as QC
+import Test.Tasty (TestTree)
+import qualified Test.Tasty.QuickCheck as TastyQC
+
+data Global d = Global { _chan :: Chan d }
+  deriving (Eq, Generic, Hashable, Show)
+
+instance Show d => Pretty (Global d) where
+  pretty = unsafeViaShow
+
+data Chan d = Chan
+  { _val :: d
+  , _rdy :: Bool
+  , _ack :: Bool
+  }
+  deriving (Eq, Generic, Hashable, Show)
+
+makeLenses ''Chan
+makeLenses ''Global
+
+instance Show d => Pretty (Chan d) where
+  pretty = unsafeViaShow
+
+data Label = Init | Send | Rcv
```

Comment:
I believe you'll need to parametrize the `Send` label on the data, so that the model checker can distinguish the different execution paths:

```suggestion
data Label d = Init | Send d | Rcv
```